cpp/src/arrow/pretty_print.cc

```diff
@@ -143,13 +145,45 @@ class ArrayPrinter : public PrettyPrinter {
   }
 
   template <typename T>
-  inline typename std::enable_if<IsInteger<T>::value, Status>::type WriteDataValues(
-      const T& array) {
+  inline typename std::enable_if<IsInteger<T>::value &&
+                                     !is_date<typename T::TypeClass>::value &&
+                                     !is_time<typename T::TypeClass>::value,
+                                 Status>::type
+  WriteDataValues(const T& array) {
     const auto data = array.raw_values();
     WriteValues(array, [&](int64_t i) { (*sink_) << static_cast<int64_t>(data[i]); });
     return Status::OK();
   }
 
+  template <typename T>
+  enable_if_date<typename T::TypeClass, Status> WriteDataValues(const T& array) {
+    const auto data = array.raw_values();
+    using unit =
+        typename std::conditional<std::is_same<T, Date32Array>::value, util::date::days,
+                                  std::chrono::milliseconds>::type;
+    WriteValues(array, [&](int64_t i) { FormatDateTime<unit>("%F", data[i], true); });
+    return Status::OK();
+  }
+
+  Status WriteDataValues(const TimestampArray& array) {
+    const int64_t* data = array.raw_values();
+    const auto type = static_cast<const TimestampType*>(array.type().get());
+    const auto fmt = "%F %T" + type->timezone();
```

Comment:
The timezone is part of the type, it shouldn't be appended to the pretty-printed values IMHO.